gh-112175: Add `eval_breaker` to `PyThreadState`

swtaarrs opened the pull request:
This change adds an `eval_breaker` field to `PyThreadState`, renaming the existing `PyInterpreterState.eval_breaker` to `instrumentation_version` (explained further down). The primary motivation is for performance in free-threaded builds: with thread-local eval breakers, we can stop a specific thread (e.g., for an async exception) without interrupting other threads.

### Pending Calls
We still want to signal all threads in an interpreter for pending calls that aren't limited to the main thread (possible since python/cpython@757b402ea1c2). These callbacks can be added by any thread, with or without the GIL held. If the targeted interpreter's GIL is locked, we signal the holding thread. When any thread is resumed, its `_PY_CALLS_TO_DO` bit is derived from the source of truth for pending calls (one of two `_pending_calls` structs). This handles situations where no thread held the GIL when the call was first added, or if the active thread did not handle the call before releasing the GIL.

In a free-threaded build, all threads in the targeted interpreter are signaled. This is slightly less efficient, both because we have to iterate over all threads to set the bit, and because it's more likely that multiple threads could call into `make_pending_calls()` before the bit is cleared. This shouldn't affect correctness, though, since `make_pending_calls()` and the `_pending_calls` structs have appropriate synchronization in place to ensure each callback happens once.

### Instrumentation Version
The source of truth for the global instrumentation version is still in `PyInterpreterState.instrumentation_version` (which used to be `eval_breaker`), in both normal and free-threaded builds. Now that this variable contains only the instrumentation version (and no more eval breaker bits), some of the code to read and write it is simplified.

Threads usually read the version from their local `eval_breaker`, where it continues to be colocated with the eval breaker bits, and the method for keeping it up to date depends on build type. Both builds first update the version in `instrumentation_version`, and then:
- Normal builds update the version in the current thread's `eval_breaker`. When a new thread takes the GIL, it copies the current version from `instrumentation_version`.
- Free-threaded builds again iterate over all threads in the current interpreter, updating the version on each one. Instrumentation (and the specializing interpreter more generally) will need more work to be compatible with free-threaded builds, so these changes are mostly intended to maintain the status quo in normal builds for now.

### Other notable changes
- The `_PY_*_BIT` macros now expand to the actual bit being set, rather than the bit's index. I think this is simpler overall.
- `_Py_set_eval_breaker_bit(interp, bit, set)` has been split into `_Py_set_eval_breaker_bit(interp, bit)` and `_Py_unset_eval_breaker_bit(interp, bit)`. This lets us use a single atomic `or`/`and`, rather than a loop around an atomic compare/exchange like the old code.

- Issue: gh-112175

swtaarrs committed:
gh-112175: Add `eval_breaker` to `PyThreadState`

This change adds an `eval_breaker` field to `PyThreadState`, renaming the existing `eval_breaker` to
`interp_eval_breaker` (its uses are explained further down). The primary motivation is for
performance in free-threaded builds: with thread-local eval breakers, we can stop a specific thread
(e.g., for an async exception) without interrupting other threads.

There are still two situations where we want the first available thread to handle a request:
- Running a garbage collection: In normal builds, we set `_PY_GC_SCHEDULED_BIT` on the current
  thread. In case a thread suspends before handling the collection, the bit is copied to and from
  `interp_eval_breaker` on thread suspend and resume, respectively. In a free-threaded build, we
  simply iterate over all threads and set the bit. The first thread to check its eval breaker runs
  the collection, unsetting the bit on all threads.
  - Free-threaded builds could have multiple threads attempt a GC from one trigger if we get very
    unlucky with thread scheduling. I didn't put any protections against this in place because a)
    the consequences of it happening are just that one or more threads will check the GC thresholds
    right after a collection finishes, which won't affect correctness and b) it's incredibly,
    vanishingly unlikely.
- Pending calls not limited to the main thread (possible since python/cpython@757b402ea1c2). This is
  a little tricker, since the callback can be added from any thread, with or without the GIL
  held. If the targeted interpreter's GIL is locked, we signal the holding thread. When a thread is
  resumed, its `_PY_CALLS_TO_DO` bit is derived from the source of truth for pending calls (one of
  two `_pending_calls` structs). This handles situations where no thread held the GIL when the call
  was first added, or if the active thread did not handle the call before releasing the GIL. In a
  free-threaded build, all threads all signaled, similar to scheduling a GC.

The source of truth for the global instrumentation version is still in `interp_eval_breaker`, in
both normal and free-threaded builds. Threads usually read the version from their local
`eval_breaker`, where it continues to be colocated with the eval breaker bits, and the method for
keeping it up to date depends on build type. All builds first update the version in
`interp_eval_breaker`, and then:
- Normal builds update the version in the current thread's `eval_breaker`. When a thread takes the
  GIL, it copies the current version from `interp_eval_breaker` as part of the same operation that
  copies `_PY_GC_SCHEDULED_BIT`.
- Free-threaded builds again iterate over all threads in the current interpreter, updating the
  version on each one.
Instrumentation (and the specializing interpreter more generally) will need more work to be
compatible with free-threaded builds, so these changes are just intended to maintain the status quo
in normal builds for now.

Other notable changes are:
- The `_PY_*_BIT` macros now expand to the actual bit being set, rather than the bit's index. I
  think this is simpler overall. I also moved their definitions from `pycore_ceval.h` to
  `pycore_pystate.h`, since their main usage is on `PyThreadState`s now.
- Most manipulations of `eval_breaker` are done with a new pair of functions:
  `_PyThreadState_Signal()` and `_PyThreadState_Unsignal()`. Having two separate functions to
  set/unset a bit, rather than one function that takes the bit value to use, lets us use a single
  atomic `or`/`and`, rather than a loop around an atomic compare/exchange like the old
  `_Py_set_eval_breaker_bit` function.

Existing tests provide pretty good coverage for most of this functionality. The one new test I added
is to make sure a GC still happens if a thread schedules it then drops the GIL before the GC runs. I
don't love how complicated this test ended up so I'm open to other ideas for how to test this (or
other things to test in general).

colesbury commented on Python/instrumentation.c:
```diff
@@ -891,18 +891,39 @@ static inline int most_significant_bit(uint8_t bits) {
-    assert((version & _PY_EVAL_EVENTS_MASK) == 0);
-    uintptr_t old = _Py_atomic_load_uintptr(&interp->ceval.eval_breaker);
+    uintptr_t old = _Py_atomic_load_uintptr(breaker);
     intptr_t new;
```
Predates your PR, but this looks like it should be `uintptr_t`:

```suggestion
    uintptr_t new;
```

colesbury commented on Python/ceval_gil.c:
```diff
@@ -956,38 +1015,38 @@ int
 _Py_HandlePending(PyThreadState *tstate)
 {
     PyInterpreterState *interp = tstate->interp;
```
variable is unused (see GHA warning)

colesbury commented on Include/internal/pycore_ceval_state.h:
```diff
@@ -78,11 +78,14 @@ struct _ceval_runtime_state {
+     * of truth when a thread is running.
+     *
+     * It is by far the hottest field in this struct and should be placed at
+     * the beginning. */
```
The "hottest field" comment may no longer accurate if `CHECK_EVAL_BREAKER()` now uses the field in PyThreadState.

swtaarrs replied:
I'll also remove the padding that says it's there to avoid false sharing. I decided against adding similar padding to `PyThreadState` since it's per-thread rather than per-interpreter, but I can be convinced otherwise.

colesbury commented on Python/ceval_gil.c:
```diff
@@ -646,11 +661,9 @@ PyEval_RestoreThread(PyThreadState *tstate)
 
 void
-_PyEval_SignalReceived(PyInterpreterState *interp)
+_PyEval_SignalReceived()
```
```suggestion
_PyEval_SignalReceived(void)
```

(or take a `_PyRuntimeState *runtime` argument)

colesbury commented on Python/ceval_gil.c:
```diff
@@ -724,8 +754,16 @@ _PyEval_AddPendingCall(PyInterpreterState *interp,
-    SIGNAL_PENDING_CALLS(interp);
+    if (main_only) {
+        _PyThreadState_Signal(_PyRuntime.main_tstate, _PY_CALLS_TO_DO_BIT);
+    } else {
```
```suggestion
    }
    else {
```

colesbury commented on Include/internal/pycore_pystate.h:
```diff
@@ -282,6 +282,42 @@ static inline _PyFreeListState* _PyFreeListState_GET(void)
+#define _PY_EVAL_EVENTS_MASK ((1U << _PY_EVAL_EVENTS_BITS)-1)
+
+static inline void
+_PyThreadState_Signal(PyThreadState *tstate, uintptr_t bit)
```
I'd suggest keeping the existing naming scheme and placement in `pycore_ceval.h` instead of `pycore_pystate.h` unless there's a concrete reason that requires moving the functions.

I think it still worth splitting setting and clearing the bits like you do here now that we have functions for atomic and & atomic or.

swtaarrs replied:
I think my only reason for the move was that your 3.12 commit that I used for reference did it (but that was before Mark's gh-109846). I'll move them back unless I come up with something more conrete.

For the naming scheme, how should the functions that operate on all threads in an interpreter fit in? Something like this?
- `_Py_set_eval_breaker_bit(PyThreadState*, uintptr_t)`
- `_Py_unset_eval_breaker_bit(PyThreadState*, uintptr_t)`
- `_Py_set_eval_breaker_bit_all(PyInterpreterState*, uintptr_t)`
- `_Py_unset_eval_breaker_bit_all(PyInterpreterState*, uintptr_t)`

Mildly related: I'd still like to figure out a way to handle thread-agnostic pending calls without signaling all threads, since that will be the only instance of that pattern once I make the GC signaling change. For the main interpreter, I can see it making sense to only signal the main thread (since that's how OS signals are handled), but what about subinterpreters? Would it be reasonable to arbitrarily pick a thread and signal just that one, assuming that any well-behaved Python thread will check its eval_breaker often enough?

colesbury commented on Python/ceval_gil.c:
```diff
@@ -702,6 +715,24 @@ _pop_pending_call(struct _pending_calls *pending,
 }
 
+static void
+signal_active_thread(PyInterpreterState *interp, uintptr_t bit)
```
Surround this with an `#ifndef Py_GIL_DISABLED` to address unused function warning

colesbury commented on Python/gc_free_threading.c:
```diff
@@ -1480,9 +1480,9 @@ PyObject_IS_GC(PyObject *obj)
+_Py_ScheduleGC(PyThreadState *tstate)
 {
-    _Py_set_eval_breaker_bit(interp, _PY_GC_SCHEDULED_BIT, 1);
+    _PyInterpreterState_SignalAll(tstate->interp, _PY_GC_SCHEDULED_BIT);
```
We only need to set the bit for the current thread.

We don't need to worry about missing the signal. If we don't run the GC first, other threads will also see that a GC needs to be scheduled and set the bit on their own eval breakers.

colesbury commented on Python/instrumentation.c:
```diff
@@ -891,18 +891,39 @@ static inline int most_significant_bit(uint8_t bits) {
+    // Set the version on all threads in free-threaded builds.
+    _PyRuntimeState *runtime = &_PyRuntime;
+    HEAD_LOCK(runtime);
+    for (; tstate; tstate = PyThreadState_Next(tstate)) {
```
This looks like it starts iterating from the current thread so it may miss some threads in the interpreter.

I think you will want to start at `interp->threads.head`

colesbury commented on Python/gc.c:
```diff
@@ -1787,13 +1789,14 @@ _PyObject_GC_Link(PyObject *op)
 void
 _Py_RunGC(PyThreadState *tstate)
 {
+    _PyThreadState_Unsignal(tstate, _PY_GC_SCHEDULED_BIT);
```
I am more comfortable with the bit clearing happening in the places that check for the bit (i.e., `_Py_HandlePending()` and in `signalmodule.c`) rather than being pushed down into this function.

The pattern that I imagine is:

1) Check for the relevant eval breaker bit
2) Clear the bit
3) Take some action

Forgetting to clear the bit leads to problems. It's easier to see that we're adhering to the pattern if checking and clearing the bit are next to each other in, for example `_Py_HandlePending()`.

swtaarrs replied:
I originally made this change because there were two different places that cleared the bit (`_Py_HandlePending()` and `PyErr_CheckSignals()`), and "clearing the bit" meant different things in normal and free-threaded builds. So I was piggybacking on the existing `gc.c`/`gc_free_threading.c` split to avoid some duplication.

But your reasoning makes sense, and especially with the change to only set the bit on the current thread in free-threaded builds, the behavioral difference goes away.

colesbury commented on Python/ceval_gil.c:
```diff
@@ -56,60 +56,78 @@
-{
-    _Py_set_eval_breaker_bit(interp, _PY_CALLS_TO_DO_BIT, 1);
-}
+update_interp_eval_breaker(PyThreadState *tstate, PyInterpreterState *interp) {
```
```suggestion
update_interp_eval_breaker(PyThreadState *tstate, PyInterpreterState *interp)
{
```

colesbury reviewed (commented):
Thanks Brett! This looks good overall, but the free-threaded GitHub CI is failing and there are a few warnings that need to be addressed. I've left comments inline as well.

### _PY_GC_SCHEDULED_BIT

For the free-threaded build, I think we only need to set the GC scheduled bit on the current thread. It doesn't matter much if we get de-scheduled -- other threads will see that a GC needs to be scheduled soon enough. Regarding "multiple threads attempt a GC": that can be a performance issue once the GIL is disabled, but will be addressed in gh-114880.

I think the same logic applies to the default build, but I'd like Mark's opinion on it. In other words, I don't think we need to copy the `_PY_GC_SCHEDULED_BIT` back to the interpreter eval breaker. In the rare case where we get swapped out between setting `_PY_GC_SCHEDULED_BIT` and running the GC, the newly scheduled thread will figure out that it needs to run the GC after the next GC allocation.

### Instrumentation

In the free-threaded build, I think we'll also need to copy the instrumentation bits when we create a new thread (while holding `HEAD_LOCK(runtime)`).

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

swtaarrs committed:
Remove unnecessary test

swtaarrs committed:
Review comments from Sam

swtaarrs committed:
Add !IsSignalled check to free-threaded _Py_ScheduleGC to match normal build

swtaarrs committed:
Move breaker bits/functions back to pycore_ceval.h, restore old naming convention

swtaarrs committed:
Merge remote-tracking branch 'upstream/main' into cpython-eval-breaker.

Minor conflicts from the new _PY_EVAL_EXPLICIT_MERGE_BIT flag.

colesbury commented on Include/internal/pycore_ceval.h:
```diff
@@ -200,40 +199,43 @@ int _PyEval_UnpackIterable(PyThreadState *tstate, PyObject *v, int argcnt, int a
 /* Reserve a few bits for future use */
 #define _PY_EVAL_EVENTS_BITS 8
-#define _PY_EVAL_EVENTS_MASK ((1 << _PY_EVAL_EVENTS_BITS)-1)
+#define _PY_EVAL_EVENTS_MASK ((1U << _PY_EVAL_EVENTS_BITS)-1)
```
I think the switch to unsigned here may lead to problems due to zero extension. For example,

```c
uintptr_t mask = ~_PY_EVAL_EVENTS_MASK;  // 0x00000000FFFFFF00
```

where as previously we would have `0xFFFFFFFFFFFFFF00`

swtaarrs replied:
I think I made this change before I fully paged in how the mask was used; I'll flip it back.

That reminds me that I meant to follow up on [this comment](https://github.com/python/cpython/blob/main/Python/instrumentation.c#L1773-L1775) from @markshannon. That comment, plus other code that operates on the version number as a `uint32_t`, suggest that the upper 32 bits of `eval_breaker` will always be 0 (on platforms where `uintptr_t` is 64 bits). Should we commit to that and make `eval_breaker` a `uint32_t`?

It wouldn't make a meaningful change to the size of the struct but could be worth doing for consistency.

colesbury replied:
Switching to `uint32_t` seems fine, but it also seems fine to keep it as `uintptr_t` in the interest of keeping the PR smaller.

I think previously (i.e., in 3.12) we only had atomic operations implemented for `int` and `uintptr_t`.

swtaarrs committed:
Back to signed 1 in _PY_EVAL_EVENTS_MASK

swtaarrs committed:
Don't copy _PY_GC_SCHEDULED_BIT to/from interp_eval_breaker

This means that the instrumentation version was the only thing left in
`interp_eval_breaker`, so I renamed it and performed some related
simplification.

swtaarrs committed:
Merge remote-tracking branch 'upstream/main' into cpython-eval-breaker

colesbury commented on Python/ceval.c:
```diff
@@ -5,7 +5,7 @@
 #include "pycore_abstract.h"      // _PyIndex_Check()
 #include "pycore_call.h"          // _PyObject_CallNoArgs()
-#include "pycore_ceval.h"         // _PyEval_SignalAsyncExc()
+#include "pycore_ceval.h"
```
Is this change intentional? Otherwise, should probably revert it.

swtaarrs replied:
It was intentional: I removed the `_PyEval_SignalAsyncExc()` function and replaced the one call to it (which wasn't even in this file) with `_Py_set_eval_breaker_bit(...)`. I tried removing the `#include` and saw that a number of other functions from `pycore_ceval.h` are still used.

colesbury commented on Python/ceval_gil.c:
```diff
@@ -56,60 +56,56 @@
-}
+    if (tstate == NULL) {
+        return;
+    }
```
We should delete this check because `tstate` can't be `NULL` here.

colesbury commented on Python/instrumentation.c:
```diff
@@ -891,18 +891,42 @@ static inline int most_significant_bit(uint8_t bits) {
 global_version(PyInterpreterState *interp)
 {
-    return interp->ceval.eval_breaker & ~_PY_EVAL_EVENTS_MASK;
+    return _Py_atomic_load_uintptr_relaxed(&interp->ceval.instrumentation_version);
```
This needs a cast to `uint32_t` now

swtaarrs replied:
I can add the cast, but I don't think this changed the type of the expression being returned - `eval_breaker` was already a `uintptr_t`, so `interp->ceval.eval_breaker & ~_PY_EVAL_EVENTS_MASK` should also be `uintptr_t`.

colesbury reviewed (commented):
Looks pretty good, a few comments below

swtaarrs committed:
Remove unnecessary NULL check, add uint32_t cast

swtaarrs committed:
Merge remote-tracking branch 'upstream/main' into cpython-eval-breaker

colesbury commented on Include/internal/pycore_ceval_state.h:
```diff
@@ -78,13 +78,10 @@ struct _ceval_runtime_state {
+    /* This variable holds the global instrumentation version. When a thread is
+       running, this value is overlaid onto PyThreadState.eval_breaker so that
+       changes in the instrumentation version will trigger the eval breaker. */
+    uintptr_t instrumentation_version;
```
Now that this is just the instrumentation_version, did you want to make this a `uint32_t` and simplify `set_version_raw`?

swtaarrs replied:
I decided to keep it as-is to keep the PR more focused. To keep things more consistent it would make sense to change the new `eval_breaker` to `uint32_t` as well, and that would make the change significantly bigger.

colesbury reviewed (commented):
This looks good to me other than the merge conflicts. Those were probably from conflicts with my PRs -- sorry about that.

Did you want to make `instrumentation_version` a `uint32_t` now that it doesn't include the eval breaker? Either way seems fine to me.

swtaarrs committed:
Merge remote-tracking branch 'upstream/main' into cpython-eval-breaker

colesbury reviewed (approved):
Thanks Brett!
